### ISSUE TITLE
Tooltips point to an incorrect definition

### ISSUE THREAD

**Issue by alphalambda** [google/codeworld#907]
**Shared Program**: https://code.world/haskell#PzVlsrPr2r_CQLbAflh8H4w

**Description**

If you hover over the `Cw.translated` in line 10, you will see that it points to the definition of `translated` in line 9, but that is not the correct definition for `Cw.translated`

The problem is that the tooltips are looking only at the name of the function and ignoring the module part.

**Comment by cdsmith** [comment]
This is a known limitation, for new.  I believe @devanshbatra04 is hoping to work on fixing this as part of Summer of Code this summer.

**Comment by cdsmith** [comment]
It looks like it's worth disabling help tooltips in haskell-mode, for the same reason that we disable autocomplete.  It's just too incomplete to be useful, and support for the full Haskell language is spotty.

**Comment by devanshbatra04** [comment]
> This is a known limitation, for now. I believe @devanshbatra04 is hoping to work on fixing this as part of Summer of Code this summer.

Yes, we should have a strategy in place once a parser is in place #863 
Should I look into a parser/lexer as well for the time?

**Comment by cdsmith** [comment]
@devanshbatra04 To be clear, we're still waiting on decisions on Summer of Code.  That will be released May 6.  So I don't know whether the project is even accepted yet.

Whether it's your summer of code project or not, though, yeah we definitely need more parsing.  It's a tricky problem, because we really need all three of (a) indentation-aware parsing, (b) superb error recovery, and (c) preferably, an incremental algorithm to fit into CodeMirror's structure because parsing gets mixed into syntax highlighting.  That's a difficult combination.

This also isn't just about parsing.  There's name resolution here, too.  Ideally, when hovering over a function, you'd recognize that the function argument shadows imported symbols of the same name, for example.  That's not easy.

**Comment by devanshbatra04** [comment]
> @devanshbatra04 To be clear, we're still waiting on decisions on Summer of Code. That will be released May 6. So I don't know whether the project is even accepted yet.

I do understand that. I just had a look at the PR and saw there were a bunch of other parsers being discussed (moo, parsimmon, chevrotain) so I thought I should chip in a bit as this seemed crucial. 

> There's name resolution here, too. Ideally, when hovering over a function, you'd recognize that the function argument shadows imported symbols of the same name, for example. 

Yes, I am clear on this front. It does seem to depend on getting the parsed symbol names for arguments and functions first.

**Comment by cdsmith** [comment]
#1175 is a special case of this.  The general case is harder, though, since top-level symbols can also be shadowed by variables with smaller scope, and I have no idea how to solve that one without reinventing GHC in the browser.
